packages/engine-core/src/binary/BinaryEngine.ts

```diff
@@ -236,36 +234,40 @@ You may have to run ${chalk.greenBright('prisma generate')} for your changes to
     this.checkForTooManyEngines()
   }
 
-  private setError(err: Error | RustLog | RustError) {
+  /**
+   * @deprecated do not rely on this
+   */
+  private setError(err: any) {
     if (isRustError(err)) {
-      this.lastRustError = err
-      this.logEmitter.emit(
-        'error',
-        new PrismaClientRustError({
-          clientVersion: this.clientVersion!,
-          error: err,
-        }),
-      )
-      if (err.is_panic) {
-        this.handlePanic()
-      }
+      this.handleRustError(err)
     } else if (isRustErrorLog(err)) {
       this.lastErrorLog = err
-      this.logEmitter.emit(
-        'error',
-        new PrismaClientRustError({
-          clientVersion: this.clientVersion!,
-          log: err,
-        }),
-      )
-      if (err.fields?.message === 'PANIC') {
-        this.handlePanic()
-      }
+      this.handleRustError(err)
     } else {
-      this.logEmitter.emit('error', err)
+      this.emitError(err)
+    }
+  }
+
+  private handleRustError(err: RustLog | RustError) {
+    const wrappedError = new PrismaClientRustError({
+      clientVersion: this.clientVersion!,
+      error: err,
+    })
+    this.emitError(wrappedError)
+    if (wrappedError.isPanic()) {
+      this.handlePanic()
     }
   }
 
+  private emitError(err: any) {
```

Comment:
(Warning: Potentially stupid question)
Why are we emitting errors here still vs always emitting in `RequestHandler`, is it for a specific set of errors that only happen in BinaryEngine? https://github.com/prisma/prisma/pull/16287/files#r1029921786

Reply:
I have simplified the binary engine error workflow extensively by relying on the type system, in particular removing the union type `RustError`, consolidating the states: `lastPanic`, `lastError`, `lastErrorlog` into `lastError: PrismaClientRustError` which allowed me to remove a lot of conditionals: and finally getting rid of the error emission, in `setError` (now it just sets the error, handling a possible panic), and making it explicit in the places where the state of the server is managed. 

https://github.com/prisma/prisma/pull/16287/commits/00a497ed5ad82deebbcdcb09dd633bfb0c00d7ce

I will still need to do some manual testing on different error conditions. I would love to be able to automate things like I with https://github.com/Shopify/toxiproxy to introduce latency or network errors as part of the test suite, but I think that goes way beyond the scope of that issue.